storm-core/src/clj/backtype/storm/stats.clj

```diff
@@ -373,4 +362,1222 @@
+                          aggregate-count-streams
+                          (merge-with + (:window->failed acc-stats)))}))
+
+(defn- agg-spout-exec-win-stats
+  "A helper function that aggregates windowed stats from one spout executor."
+  [acc-stats new-stats include-sys?]
+  (let [{w->compLatWgtAvg :completeLatencyTotal
+         w->acked :acked}
+          (swap-map-order
+            (into {} (for [w (keys (:acked new-stats))]
+                       [w (agg-spout-lat-and-count
+                            (get (:complete-latencies new-stats) w)
+                            (get (:acked new-stats) w))])))
+        handle-sys-components-fn (mk-include-sys-filter include-sys?)]
+    {:window->emitted (->> (:emitted new-stats)
+                           (map-val handle-sys-components-fn)
+                           aggregate-count-streams
+                           (merge-with + (:window->emitted acc-stats)))
+     :window->transferred (->> (:transferred new-stats)
+                               (map-val handle-sys-components-fn)
+                               aggregate-count-streams
+                               (merge-with + (:window->transferred acc-stats)))
+     :window->comp-lat-wgt-avg (merge-with +
+                                           (:window->comp-lat-wgt-avg acc-stats)
+                                           w->compLatWgtAvg)
+     :window->acked (->> (:acked new-stats)
+                         aggregate-count-streams
+                         (merge-with + (:window->acked acc-stats)))
+     :window->failed (->> (:failed new-stats)
+                          aggregate-count-streams
+                          (merge-with + (:window->failed acc-stats)))}))
+
+(defmulti agg-comp-exec-stats
+  "Combines the aggregate stats of one executor with the given map, selecting
+  the appropriate window and including system components as specified."
+  (fn dispatch-fn [_ _ init-val _] (:type init-val)))
+
+(defmethod agg-comp-exec-stats :bolt
+  [window include-sys? acc-stats new-data]
+  (assoc (agg-bolt-exec-win-stats acc-stats (:stats new-data) include-sys?)
+         :stats (merge-agg-comp-stats-comp-page-bolt
+                  (:stats acc-stats)
+                  (agg-pre-merge-comp-page-bolt new-data window include-sys?))
+         :type :bolt))
+
+(defmethod agg-comp-exec-stats :spout
+  [window include-sys? acc-stats new-data]
+  (assoc (agg-spout-exec-win-stats acc-stats (:stats new-data) include-sys?)
+         :stats (merge-agg-comp-stats-comp-page-spout
+                  (:stats acc-stats)
+                  (agg-pre-merge-comp-page-spout new-data window include-sys?))
+         :type :spout))
+
+(defn- aggregate-comp-stats*
+  [window include-sys? data init-val]
+  (-> (partial agg-comp-exec-stats
+               window
+               include-sys?)
+      (reduce init-val data)))
+
+(defmulti aggregate-comp-stats
+  (fn dispatch-fn [& args] (-> args last first :stats :type)))
```

Comment:
This could result in a null dispatch value if the component's heartbeat does not contain metrics.  We should be using the `:type` directly instead of accessing it via `:stats`.